pkg/tls_extensions/atlsLIstener.go

```diff
@@ -0,0 +1,397 @@
+}
+
+//export validationVerificationCallback
+func validationVerificationCallback(teeType C.int) uintptr {
+	switch int(teeType) {
+	case NoTee:
+		return uintptr(0)
+	case AmdSevSnp:
+		return registerValidationVerification(attest.VerifyAttestationReportTLS)
+	default:
+		return uintptr(0)
+	}
+}
+
+//export fetchAttestationCallback
+func fetchAttestationCallback(teeType C.int) uintptr {
+	switch int(teeType) {
+	case NoTee:
+		return uintptr(0)
+	case AmdSevSnp:
+		return registerFetchAttestation(attest.FetchAttestation)
+	default:
+		return uintptr(0)
+	}
+}
+
+//export callVerificationValidationCallback
+func callVerificationValidationCallback(callbackHandle uintptr, attReport *C.uchar, attReportSize C.int, repData *C.uchar) C.int {
+	handle := cgo.Handle(callbackHandle)
+	defer handle.Delete()
+
+	callback := handle.Value().(ValidationVerification)
+	attestationReport := C.GoBytes(unsafe.Pointer(attReport), attReportSize)
+	reportData := C.GoBytes(unsafe.Pointer(repData), agent.ReportDataSize)
+
+	return C.int(callback(attestationReport, reportData))
+}
+
+//export callFetchAttestationCallback
+func callFetchAttestationCallback(callbackHandle uintptr, reportDataByte *C.uchar, outlen *C.int) *C.uchar {
+	handle := cgo.Handle(callbackHandle)
+	defer handle.Delete()
+
+	callback := handle.Value().(FetchAttestation)
+	reportData := C.GoBytes(unsafe.Pointer(reportDataByte), agent.ReportDataSize)
+
+	res := callback(reportData)
+	if res == nil {
+		fmt.Fprintf(os.Stderr, "attestation callback returned nill")
+		return nil
+	}
+
+	*outlen = C.int(len(res))
+	resultC := C.malloc(C.size_t(len(res)))
+	if resultC == nil {
+		fmt.Fprintf(os.Stderr, "could not allocate memory for fetch attestation callback")
+		return nil
+	}
+
+	C.memcpy(resultC, unsafe.Pointer(&res[0]), C.size_t(len(res)))
+
+	return (*C.uchar)(resultC)
+}
+
+type ATLSServerListener struct {
+	tlsListener *C.tls_server_connection
+}
+
+func Listen(addr string, cert []byte, key []byte) (net.Listener, error) {
+	ip, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, fmt.Errorf("error while creating listener: %v", err)
+	}
+
+	p, err := strconv.Atoi(port)
+	if err != nil {
+		return nil, fmt.Errorf("bad format of IP address: %v", err)
+	}
+
+	cCertPEM := (*C.char)(unsafe.Pointer(&cert[0]))
+	cKeyPEM := (*C.char)(unsafe.Pointer(&key[0]))
+	cIP := C.CString(ip)
+	defer C.free(unsafe.Pointer(cIP))
+
+	atlsListener := C.start_tls_server(
+		cCertPEM, C.int(len(cert)),
+		cKeyPEM, C.int(len(key)),
+		cIP, C.int(p))
+	if atlsListener == nil {
+		return nil, fmt.Errorf("could not create listener")
+	}
+
+	return &ATLSServerListener{tlsListener: atlsListener}, nil
+}
+
+// accept implements the Accept method in the [Listener] interface; it
+// waits for the next call and returns a generic [Conn].
+func (l *ATLSServerListener) Accept() (net.Conn, error) {
+	conn := C.tls_server_accept(l.tlsListener)
+	if conn == nil {
+		return &ATLSConn{tlsConn: nil}, nil
```

Comment:
if conn is nil, return an error

Reply:
If an error is returned, the Agent server will stop. GRPC will stop it. To avoid that and stop the connection when the attestation verification fails, the return value is ATLSConn object with Conn of value nil.